test/e2e/storage/testsuites/snapshottable_metrics.go

```diff
@@ -0,0 +1,400 @@
+
+				ginkgo.By("checking for CreateSnapshotAndReady metrics")
+				err = storageutils.WaitForSnapshotReady(dc, vs.GetNamespace(), vs.GetName(), framework.Poll, f.Timeouts.SnapshotCreate)
+				createSnapshotAndReadyMetrics.waitForSnapshotControllerMetric(originalCreateSnapshotAndReadyCount+1.0, f.Timeouts.SnapshotControllerMetrics)
+
+				ginkgo.By("delete and wait for complete")
+				err = storageutils.DeleteAndWaitSnapshot(dc, vs.GetNamespace(), vs.GetName(), framework.Poll, f.Timeouts.SnapshotDelete)
+				framework.ExpectNoError(err)
+
+				ginkgo.By("check for delete metrics")
+				metricsTestConfig.operationName = "DeleteSnapshot"
+				deleteSnapshotMetrics.waitForSnapshotControllerMetric(originalDeleteSnapshotCount+1.0, f.Timeouts.SnapshotControllerMetrics)
+			})
+		})
+	})
+}
+
+type snapshotMetricsTestConfig struct {
+	// expected values
+	metricName      string
+	metricType      string
+	driverName      string
+	operationName   string
+	operationStatus string
+	snapshotType    string
+	le              string
+}
+
+type snapshotControllerMetrics struct {
+	// configuration for metric
+	cfg            snapshotMetricsTestConfig
+	metricsGrabber *e2emetrics.Grabber
+
+	// results
+	countMetrics  map[string]float64
+	sumMetrics    map[string]float64
+	bucketMetrics map[string]float64
+}
+
+func newSnapshotMetricsTestConfig(metricName, metricType, driverName, operationName, operationStatus, le string, pattern storageframework.TestPattern) snapshotMetricsTestConfig {
+	var snapshotType string
+	switch pattern.SnapshotType {
+	case storageframework.DynamicCreatedSnapshot:
+		snapshotType = "dynamic"
+
+	case storageframework.PreprovisionedCreatedSnapshot:
+		snapshotType = "pre-provisioned"
+
+	default:
+		framework.Failf("invalid snapshotType: %v", pattern.SnapshotType)
+	}
+
+	return snapshotMetricsTestConfig{
+		metricName:      metricName,
+		metricType:      metricType,
+		driverName:      driverName,
+		operationName:   operationName,
+		operationStatus: operationStatus,
+		snapshotType:    snapshotType,
+		le:              le,
+	}
+}
+
+func newSnapshotControllerMetrics(cfg snapshotMetricsTestConfig, metricsGrabber *e2emetrics.Grabber) *snapshotControllerMetrics {
+	return &snapshotControllerMetrics{
+		cfg:            cfg,
+		metricsGrabber: metricsGrabber,
+
+		countMetrics:  make(map[string]float64),
+		sumMetrics:    make(map[string]float64),
+		bucketMetrics: make(map[string]float64),
+	}
+}
+
+func (scm *snapshotControllerMetrics) waitForSnapshotControllerMetric(expectedValue float64, timeout time.Duration) {
+	metricKey := scm.getMetricKey()
+	if successful := utils.WaitUntil(10*time.Second, timeout, func() bool {
+		// get metric value
+		actualValue, err := scm.getSnapshotControllerMetricValue()
+		if err != nil {
+			framework.Logf("failed to find metric value: %v", err)
+			return false
+		}
+
+		// Another operation could have finished from a previous test,
```

Comment:
If you run the test with [Serial], then you reduce the chance of metrics from another test conflicting (unless a previous test failed). We'll need to add a new test job for snapshot serial tests though.

Reply:
I can add this in afterwards once this PR is merged.